packages/ra-core/src/controller/useListController.ts

```diff
@@ -184,15 +184,31 @@ const useListController = <RecordType = Record>(
 
     const finalIds = typeof total === 'undefined' ? defaultIds : ids;
 
+    const totalPages = useMemo(() => {
+        return Math.ceil(total / query.perPage) || 1;
+    }, [query, total]);
+
     useEffect(() => {
         if (
             query.page <= 0 ||
             (!loading && query.page > 1 && (finalIds || []).length === 0)
         ) {
-            // query for a page that doesn't exist, set page to 1
+            // Query for a page that doesn't exist, set page to 1
             queryModifiers.setPage(1);
+        } else if (!loading && query.page > totalPages) {
+            // Query for a page out of bounds, set page to the last existing page
+            // It occurs when deleting the last element of the last page
+            queryModifiers.setPage(totalPages);
         }
-    }, [loading, query.page, finalIds, queryModifiers, total, defaultIds]);
+    }, [
+        loading,
+        query,
```

Comment:
why changing query.page to query here ? only query.page is used in the useEffect